app/ydoc-server/src/languageServerSession.ts

```diff
@@ -570,7 +587,23 @@ class ModulePersistence extends ObservableV2<{ removed: () => void }> {
           syncModule.applyEdit(edit)
         } else {
           const { root, spans } = Ast.parseModuleWithSpans(code, syncModule)
-          syncModule.syncRoot(root)
+
+          let rootToSync = root
+          const metadataSnapshot = metadata.ide.snapshot
+          if (metadataSnapshot && idMapJson) {
+            const snapshotCode = ModulePersistence.decodeCodeSnapshot(metadataSnapshot)
+            if (snapshotCode !== code || 1 === 1 /* simulate code mismatch */) {
+              const { root, spans } = Ast.parseModuleWithSpans(snapshotCode)
+              const edit = root.module.edit()
+              Ast.setExternalIds(edit, spans, deserializeIdMap(idMapJson))
+              edit.getVersion(root).syncToCode(code)
+              root.module.applyEdit(edit)
+
+              rootToSync = root
+            }
+          }
+
+          syncModule.syncRoot(rootToSync)
```

Comment:
The `syncRoot` call needs to occur before the `syncToCode`; it can be done right after parsing (the name might be misleading; it sets the root object, but changes to the object that occur later will still be seen).